compiler/rustc_hir_analysis/src/astconv/mod.rs

```diff
@@ -2098,15 +2095,157 @@ impl<'o, 'tcx> dyn AstConv<'tcx> + 'o {
+            let infcx = tcx.infer_ctxt().ignoring_regions().build();
+            let param_env = tcx.param_env(impl_);
+
+            let impl_ty = tcx.bound_type_of(impl_);
+            let impl_substs = self.fresh_item_substs(impl_, &infcx);
+            let impl_ty = impl_ty.subst(tcx, impl_substs);
+
+            let InferOk { value: impl_ty, obligations } =
+                infcx.at(&cause, param_env).normalize(impl_ty);
+
+            // Check that the Self-types can be related.
+            let Ok(InferOk { obligations: sub_obligations, value: () }) = infcx
+                .at(&ObligationCause::dummy(), param_env)
+                .define_opaque_types(false)
+                .sup(impl_ty, self_ty)
+            else {
+                continue;
+            };
+
+            // Check whether the impl imposes obligations we have to worry about.
+            let impl_bounds = tcx.predicates_of(impl_);
+            let impl_bounds = impl_bounds.instantiate(tcx, impl_substs);
+
+            let InferOk { value: impl_bounds, obligations: norm_obligations } =
+                infcx.at(&cause, param_env).normalize(impl_bounds);
+
+            let impl_obligations =
+                traits::predicates_for_generics(|_, _| cause.clone(), param_env, impl_bounds);
+
+            let candidate_obligations = impl_obligations
+                .chain(norm_obligations.into_iter())
+                .chain(obligations.iter().cloned());
+
+            let mut matches = true;
+
+            // Evaluate those obligations to see if they might possibly hold.
+            for o in candidate_obligations {
+                let o = infcx.resolve_vars_if_possible(o);
+                if !infcx.predicate_may_hold(&o) {
+                    matches = false;
+                    unsatisfied_predicates.push(o.predicate);
+                }
+            }
+
+            // Evaluate those obligations to see if they might possibly hold.
+            for o in sub_obligations {
+                let o = infcx.resolve_vars_if_possible(o);
+                if !infcx.predicate_may_hold(&o) {
+                    matches = false;
+                    unsatisfied_predicates.push(o.predicate);
+                }
+            }
+
+            if !matches {
+                continue;
+            }
+
+            self.check_assoc_ty(assoc_item, name, def_scope, block, span);
+
+            let ty::Adt(_, adt_substs) = self_ty.kind() else {
+                // FIXME(inherent_associated_types)
+                bug!("unimplemented: non-adt self of inherent assoc ty");
+            };
+
+            let item_substs =
+                self.create_substs_for_associated_item(span, assoc_item, segment, adt_substs);
+            // FIXME(inherent_associated_types): Check if the obligations arising from the
+            // where-clause & the bounds on the associated type and its parameters hold.
+            let ty = tcx.bound_type_of(assoc_item).subst(tcx, item_substs);
```

Comment:
I wonder if astconv should turn this into a `ty::Projection` with the impl item def id, then let normalization handle projecting it onto the real type and confirming the GAT substitutions. It would fix this comment:

> Check if the obligations arising from the where-clause & the bounds on the associated type and its parameters hold.

Reply:
That FIXME now has its own issue btw, namely #106722, and I can *try* to implement it returning a type of kind `TyKind::Alias(AliasKind::Projection, AliasTy { … })`. I faintly remember trying this before and it lead to ICEs down the line where the compiler wasn't happy that *inherent* assoc ty projections obviously lacked a trait reference. The docs of [`AliasTy.def_id`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/ty/sty/struct.AliasTy.html#structfield.def_id) kind of confirm this: “The `DefId` of the `TraitItem` […] if this is a projection”. Let's see if I can find another way. I wonder if rust-lang/compiler-team#504 is relevant here.

Should I attempt this in this PR though, I wonder? I want to keep things “minimal” to make life easier for reviewers and since I don't even know if the general disposition for this PR is merging or closing as the implementation is not quite clean yet.

Reply:
Do you think we are in need of a new `AliasKind::InherentProjection` (modulo naming)? I am not deeply familiar with the inner workings of the type system yet but as far as I understand at some point in the future we do need to map polymorphic *inherent* associated types to *placeholder projection types* for lazy normalization (?) & generally proper unification for IATs and to finally pave the way for their long-awaited stabilization. Do I get that right?

Reply:
> Do you think we are in need of a new AliasKind::InherentProjection (modulo naming)?

I am currently investigating this approach.

Reply:
I think this would be good as a followup.